### ISSUE TITLE
SMTCOMP

### ISSUE THREAD

**Issue by TrevorHansen** [stp/stp#287]
If STP is submitted practically unchanged, perhaps with a novel SAT solver, to SMTCOMP. Does anyone think we should ask for anything, or be involved in some way, beyond just an acknowledgement in the system description that STP is used?

The SMTCOMP rules already have the notion of a "wrapper tool", which covers the need to acknowledge:
_Wrapper tools. A wrapper tool is defined as any solver which calls one or more SMT solvers
not written by the author of the wrapper tool. The other solvers are called the wrapped solvers. A
wrapper tool must explicitly acknowledge any solvers that it wraps. Its system description should
make clear the technical innovations by which the wrapper tool expects to improve on the wrapped
solvers._

**Comment by msoos** [comment]
What novel SAT Solver? :) Norbert wrote to me that he wants to submit with his SAT solver, Riss. I'm happy for that but if so, he probably needs to add an issue here, so we can track it, right? I think it's important to be transparent and I urged him to create a pull request.

By the way, CMS has been significantly improved and you should see massive speedup compared to last year(s). So it might be an interesting challenge to see if CMS or Riss would be the best to use -- we should submit with both. I am all-in for some friendly competition, but for that to happen we need to have this tracked under a GitHub issue.

**Comment by conp-solutions** [comment]
As Mate already said, I'm eager to see whether Riss can improve the performance of STP on the one hand, and I'd love to see how Riss performs as a SMT backend in general. The source of the solver is hosted here: https://github.com/conp-solutions/riss

I can create the pull request and make Riss another SAT backend in the list of backends, I just want to know what's a reasonable way. Concerning the submission: I'm happy to go the solver wrapper way, or anything else. As Mate, I'm all for submitting more versions. Honestly, I do not know whether Riss performs better than the currently used solvers in the SMT context.

That being said: can you shed some light on how the multi-threaded version of STP works? Is it "just" replacing the sequential SAT backend with a parallel one, or do you do more?

**Comment by TrevorHansen** [comment]
Hi Norbert,

You're welcome to integrate Riss with STP. In the "SAT" directory are examples of how the other SAT solvers integrate. If it doesn't make sense, let me know and we can discuss.

STP isn't multi-threaded itself, it just calls a multi-threaded SAT solver.

If you'd like to evaluate Riss, dumping a CNF from STP seems the easiest to me. As well as being easier to implement, it might be faster too. Integrating with STP seems like it fragments memory or some-such leading to slower solving. Something i've been meaning to investigate.

**Comment by conp-solutions** [comment]
I see. I submitted STP generated CNFs to a previous SAT competition (I do not recall the exact year right now), and actually did not follow up on the results of the competition.

Did we reach a conclusion how/whether to submit (multiple versions) to the competition? That's actually a part of the motivation for integrating Riss as a backend ;)

**Comment by msoos** [comment]
If you do a pull request as per Trevor's description above, we are happy to submit your version as well as the CMS(and MiniSat possibly) versions, so basically we do the heavy-lifting for you (static binaries, compilation issues on StarExec, etc). Otherwise, you are welcome to submit your version, we will submit ours -- there is no need to coordinate in terms of submission limits, they are quite lax. If you don't do the the pull request in time, the only thing to coordinate on is that you most likely want to latest STP to submit -- on that you have to watch the github and we must agree on a tag that we will use e.g. "satcomp18" so we all submit with the same STP version to the competition. @TrevorHansen do you think we could agree on a tag like that?

Also, may I ask why you are writing under "conp-solutions"? Is that a registered/future company name? Is this for advertising?

**Comment by conp-solutions** [comment]
I am fine with both ways, in the end it comes down to when I am able to integrate Riss.

Concerning "conp-solutions", I do not produce much constraint reasoning code during my full time job (any more), but am allowed to do so in my spare time. As I still submit systems, or sometimes present talks, I needed a name (for 'affiliation', where you used 'hobbyist' in the past). The conp-solutions name popped up after a while. There is no company behind, I own a domain based on it (which does not point anywhere at the moment, priorities again).

**Comment by conp-solutions** [comment]
The pull request https://github.com/stp/stp/pull/289 is merged. I wonder how the next steps towards the SMT competition are. I'd love to understand whether I should take care of my own stp version "stp+riss" now, or whether you'll take care of it. Please tell me whether there is more to be done. Thanks!

**Comment by conp-solutions** [comment]
Sure, I can take care of Riss. I wonder whether the files that have been submitted last year are available. Going from there should be simpler than recreating everything.

Furthermore, should we agree on a commit to start from. After merging my pull request, the difference should basically come down to running stp with "--minisat", "--cryptominisat" or "--riss", and the rest of the submission should be the same - or do I miss anything. I'd steal the build instructions from the travis scripts.

**Comment by TrevorHansen** [comment]
I've edited the run scripts:
* The test job the competition organisers ran seemed stricter than before, and parsed stderr too, so I've redirected the solvers' stderr.
* The command line arguments to riss didn't seem right.

It seems to me that we need to resubmit, do you have time @conp-solutions ?

Trev.

**Comment by TrevorHansen** [comment]
Hi @msoos, 

In the buildscript CMS is being compiled with LARGEMEM=OFF. We have 61440MB available in the competition. Two questions:
1) Do we need LARGEMEM? Searching quickly in git it looks like LARGEMEM is only for > 1TB?
2) Is there some option we should give to CMS so that it can drop learnt claues (say) as it approaches the hard memory-limit, or is there something automatically done? 

Thanks!

**Comment by conp-solutions** [comment]
Yes, i can resubmit tonight with a different set of scripts being used, keeping the binaries. Luckily, updating the run scripts can be done without reuploading the solver.In case you want to have more control, i can also grant you access to the starexec space where the solvers currently reside.Please note, my responsiveness will decrease starting tomorrow, due to travelling.Best,Norbert

**Comment by msoos** [comment]
As for largemem, I don't think we need that, unless memory-out has been an issue. It's useful in case we need more than 16GB of memory to store non-binary clauses (binary clauses are stored elsewhere). It decreases performance by about 10% on all instances. On SAT competition instances, this issue never happens, it typically only happens on problems that are ridiculously large for some strange reason. I don't expect that to trigger during the competition, but I haven't worked enough with SMT problems from the competition to know exactly. I would go for not using largemem. That's my hunch.

**Comment by conp-solutions** [comment]
Considering the solver style of Riss, the ZIP file I uploaded uses a statically linked binary. Hence, I will update the configurations of the solvers only, and will not update the actual solver links. I will add the two of you to the emails I write to Tjark, so that you have visibility.

**Comment by msoos** [comment]
Thanks!

Mate

**Comment by TrevorHansen** [comment]
FYI. The smtcomp 2018 qf_bv run has finished:
https://www.starexec.org/starexec/secure/details/job.jsp?id=28689

We did about the same as last year:
http://smtcomp.sourceforge.net/2017/results-QF_BV.shtml?v=1500632282

**Comment by msoos** [comment]
Also, congrats, MinkeyRink got really high! :)

**Comment by conp-solutions** [comment]
I had a brief look at the summary and noticed Riss does not perform well on the 20 minute timeout benchmark. For 5000 seconds (SAT competition) that should be better. I can try to come up with a better 20minute-STP configuration for Riss until next year.Concerning the differences to CMS: we should think about an actual SAT portfolio for next year, mixing parallel CMS with a single core Riss. As i am currently traveling, I cannot tell whether that would result in a better performance, given the data we have. I think this is worth a shot.

**Comment by msoos** [comment]
Sounds reasonable :) I think by next year I'm moving to SMT though... The
5.6.3 CMS is probably the last release before a major release that should
change "little" or "everything" depending on how one looks at it, 5.7
probably.

So by next year I might fork STP. It will be an open system though, and I
plan not to compete with Minkeyrink or Boolector, just roll my own crap and
play. Maybe we can merge it back one day. Will probably call it STP-related
to give credit to Trevor and Vijay.

**Comment by TrevorHansen** [comment]
> What SAT solver did Yices use? 
From memory, Yices has its own built-in SAT solver.

> And what is master-2018-06-10-b19c840-competition-default -- who wrote it and what SAT solver does it use?
It's CVC4, it looks like it uses Cadical for QF_BV.

I'll run some tests with different SAT solvers and see how much difference it makes. 

Often if a SAT solver is slower on a set of problems, for example, Riss solves 310 less Bruttomesso problems than CMS, it's because STP doesn't adequately simplify at the bit-vector level. The Bruttomesso problems are crafted to show the advantage of a particular simplification.

**Comment by conp-solutions** [comment]
Riss is not really parallel (and I never really put a lot of effort into that). Therefore, given N threads, I'd assign: N-2 CMS, 1 Riss, 1 MiniSat. I did not have a look into the data whether Riss or MiniSat actually contribute to the overall result in case the parallel CMS is used. We might be able to drop one of the two to give CMS more resources.

I'm not sure about the memory barrier. One could start each solver with a memory limit and then kill them, or could limit the overall combination and the solvers die as they approach the limit (that would have to be tested). Otherwise, we could always reduce the number of cores for CMS further.

There is another change in this years competition: they ask for printing models. The current competition STP stops after evaluating the SAT result. Is there code that can turn a SAT model back into an SMT model already?

**Comment by msoos** [comment]
Oh damn, getting back to the model may be a real pain... Currently it's running as follows: STP->CNF->depending on SAT/UNSAT we print the right thing. But no model as far as I understand... :S

**Comment by conp-solutions** [comment]
Yeah, i know. I wonder whether there has been code to parse a model back from SAT, and more interestingly converts it back to SMT.Another ugly hack would be to add a solver backend that is called exactly once and calls a user specifyable binary, wrapped in a competition compilation guard. For the other backends printing the SMT model is already implemented, so at least that part is solved. And the SMT formula has to be parsed only once.Best,Norbert

**Comment by TrevorHansen** [comment]
I haven't seen the new requirements. But, SMTLIB has a print-model command
which STP mostly implements, so I imagine we'll be OK with just a little
bit of work (if any). Trev.

**Comment by msoos** [comment]
@TrevorHansen  Okay, sounds good, thanks! I think we still have time to submit our solver, the deadline is not yet up, and I can't see the setup (memory, timeout, etc) up yet. Let's wait until that is up so we can submit knowing what the conditions are. It may affect our strategy as far as I can see, right?

**Comment by TrevorHansen** [comment]
I think SMTCOMP submissions are usually in May, so we've some time.

If we're going to consider a portfolio solver, is it worth trying to
exchange length 1 (i.e. level-0 assignments) and length 2 learned clauses
between cms and riss? Is this something useful? Hard to implement? or
something that's already standard?

Trev.

**Comment by conp-solutions** [comment]
I have a simple patch that would allow to run the 3 solvers we currently have in parallel, similarly to what we did in the competition.As long as the solvers are called as external binaries, sharing is hard. Given the simplification techniques riss implements, sharing with another solver that does similar simplifications might even become unsound.I would go with the plain portfolio approach for now, and give all threads except 2 to cms, and 1 for riss and minisat respectively (except we have data that shows that cms or riss solve anything that could be solved by minisat).Best,Norbert

**Comment by msoos** [comment]
I agree, let's go for the plain portfolio approach. Just to note again, we should explore how many threads we want to have in total, so as not to run out of memory -- but I agree with Norbert on this point as well :)

Cheers,

Mate

**Comment by conp-solutions** [comment]
I just received the final call for solvers for this years competition. So, I implemented the portfolio approach, and now actually have one more solver I would like to add there, my MergeSAT solver (a fork of last years SAT winner).
Independently, are there other things that need to go into STP before it can be submitted to the competition?

**Comment by msoos** [comment]
For the moment, we only need to submit the preliminary solver. Anything that needs to go into STP can go into it later, for the moment we need to decide on the tactic we do for submitting. We can submit multiple versions:

"Multiple versions.The intent of the organizers is to promote as wide a comparison among solvers and
solver options as possible.  However, if the number of solver submissions is too large for the
computational resources available to the competition, the organizers reserve the right not to accept
multiple versions of solvers from the same solver team."

I couldn't find the part about multi-threaded solvers, though... are they not running it this year? I am a bit confused. So, we should think about how many sequential and how many multithreaded we want to submit. I suggest we submit:

- 1 single-threaded STP with MiniSat -- Could be skipped?
- 1 single-threaded STP with CMS
- 1 single-threaded STP with Riss or MergeSAT? -- Or you want one for each? Maybe we should definitely skip with MiniSat.
- 1 multi-threaded STP with MergeSAT, CMS, Riss all together

What do you all think? Or should we have 2 multi-threaded? We have to decide in a week as we need to submit all of these as preliminary solvers that we can update later and submit the PDF description also later.

For stuff that needs to go into STP itself -- @TrevorHansen  have you had a chance to checking the new rules: https://smt-comp.github.io/2019/rules19.pdf The SMT-LIB has also been updated accordingly: https://clc-gitlab.cs.uiowa.edu:2443/SMT-LIB-benchmarks/QF_BV It would be really nice if you could take a look! Thanks so much!

Looking forward to competing at the competition! CryptoMiniSat has evolved significantly and hopefully it will solve a lot more problems :)

**Comment by conp-solutions** [comment]
In the sequential case, STP+Riss nor STP+MiniSat did not achieve much (wrt ranking) last year, so that has a lower priority. For the SAT competition setup, MergeSAT should be much better, hence, I vote for the sequential cominations STP+MergeSAT and STP+CMS. I am not sure how much SMT Competition cares about unique solver contributions, and how much MiniSat or Riss could actually contribute. Riss has at least more powerful simplifications.

For the multicore case, I would love to see a portfolio STP with all 4 backends (CMS supports multi-threading, MergeSAT is recent with limited simplification, Riss has more simplification, MiniSat is MiniSat), and I can extend the multisolver tool I created recently accordingly. If you want to have more combinations, feel free to suggest!

**Comment by msoos** [comment]
Hey! Thanks for the quick answer! So if I understand correctly, you'd like to see:

- Sequential: STP + MergeSAT
- Sequential: STP + CMS
- Multithreaded: STP + MergeSAT+CMS (maybe multi-threaded?)+Riss+MiniSat

I like this setup, but I would remove MiniSat from the mix. I don't mind either way, though, it's fine by me. So I think it's good to go. Can you do the preliminary submission of all 3 them to StarExec? (it doesn't need to be perfect, we just need a placeholder, basically, though it would be nice if you could do the compilation, too). We will then need to write a 2-page description of all 3, which will be the harder part, but we can do that later. I think we should do the kind of "STP and friends at the SMT Competition" thing that Armin has been doing for the SAT Competition, where it's a single 2-3 page description, of all 3. Would that work?

@TrevorHansen  what do you think? Would the above setup work for you? Let us know!

**Comment by msoos** [comment]
As per suggestion by Norbert -- submitted set of solvers:

By Norbert:
https://www.starexec.org/starexec/secure/details/solver.jsp?id=19968
This contains:
* MergeSAT sequential non-incremental
* Riss sequential non-incremental
* CMS sequential non-incremental
* MiniSat sequential non-incremental
(i.e. all configs from stp-starexec.sh: at
https://github.com/stp/stp/blob/aa6b57f32da2e0e36b9eec2d3a2467d335d49f63/scripts/make-starexec.sh )

By Mate:
https://www.starexec.org/starexec/secure/details/solver.jsp?id=23396
This contains:
* CMS incremental (no need for the shell script here)


We perhaps are still missing:
* Portfolio (MergeSAT, (parallel) CMS, Riss, MiniSat)
* Multithreaded ((parallel) CMS)

**Comment by msoos** [comment]
I have now submitted one called "stp-mt" that will be the parallel -- we can make this the portfolio too, I'm happy either way:

https://www.starexec.org/starexec/secure/details/solver.jsp?id=23397

**Comment by conp-solutions** [comment]
I just checked the description again, and it looks like we need to submit (many) more solvers to get all configurations we want. I will drop configurations from my initial submission accordingly, and submit all the other STP configurations that have been included. The full list would then be:

Name Submitter Variant Link
"stp-incremental" Mate pure-binary https://www.starexec.org/starexec/secure/details/solver.jsp?id=23396
stp-mt Mate "parallel CMS" https://www.starexec.org/starexec/secure/details/solver.jsp?id=23397
STP Norbert "stp+external CMS" https://www.starexec.org/starexec/secure/details/solver.jsp?id=19968
stp-riss Norbert "stp+external Riss" to-be-sumitted
stp-mergesat Norbert "stp+external MergeSAT" to-be-sumitted
stp-minisat Norbert "stp+external MiniSat" to-be-sumitted
stp-portfolio Norbert "stp+external SAT Portfolio CMS[n-3], MergeSAT[1], Riss[1], MiniSat[1]" to-be-sumitted

I will submit the other variants of stp now, basically deleting all but one configuration from that package, and creating a google form for each of them. Finally, I will update this entry with the complete list, and write an email to the organizers with the priorities of the solvers, so that they know what to drop in case they have too many submissions.

**Comment by conp-solutions** [comment]
Now that all solvers have been submitted, how do we handle the system description? There's a bunch of questions I have:
Do we submit the one from last year, and updating the description of the submitted variants?
Who will submit the description?
Who (including affiliations) should be on the authors list?

**Comment by msoos** [comment]
Crap, I haven't had time to work on this at all. I feel really bad. Can you please send in the description from last year, with minor updates? I haven't worked on STP at all, only improvement from my side is newer CryptoMiniSat, as far as I'm concerned, that's it.

Also, what should we be submitting? I feel completely inadequate to do this :S I feel super-behind and I am not sure sure I will be able to handle this :S 

I was supposed to be responsible for the STP-incremental and STP-multithreaded. But I am not sure I should be doing this. I am sorry to drop the ball so late. Can someone send me a ZIP I need to upload? I got a response from the organisers that the STP I submitted was having trouble with things so I am kinda lost, it's been a while I worked on STP. Can someone pick this up for me or help me with submitting the right ZIP?  Sorry :(

Mate

**Comment by conp-solutions** [comment]
I can update and send the description (and the seed integer 1094795585. I am of little help when it comes to submitting the incremental version, and to decide in which tracks that should actually participate.

IIRC, I already sent a multi-threaded vartiant, so that should be fine.

**Comment by msoos** [comment]
Hi,

Thanks for all of that! Yeah, the integer is fine and thanks so much for
helping with the description. Glad the multithread is being handled too! I
owe you one :) Trevor -- what are we supposed to do for the incremental
version? Is there some option or flag I need to build STP with for that?

Thanks,

Mate

**Comment by TrevorHansen** [comment]
I'm fixing the model generation now. It should be ready in a couple of
hours.

From memory, incremental sends the problems via STDIN rather than
specifying a file as a command line argument. If STP uses a built-in SAT
solver, it should already work, but I'll verify.

Trev.

**Comment by msoos** [comment]
Thank you so much, let me know how it went! If you can generate the ZIP
that'd be great, otherwise, I'm happy to do so myself from latest STP and
latest `master` CryptoMiniSat.

Thanks again,

Mate

**Comment by TrevorHansen** [comment]
The model generation is now working as per the SMTLIB2 specification. So it's OK to enter the model-validation track at SMTCOMP 2019.

I'll check the incremental, too.

**Comment by TrevorHansen** [comment]
OK. The incremental needed fixing, too. It's OK to enter into the incremental QF_BV track, assuming you're using an inbuilt SAT solver

**Comment by msoos** [comment]
Thank you so much! Can you perhaps give me some example inputs/outputs
(e.g. files) just so that I can validate that all is good? Thank you again,

Mate

**Comment by conp-solutions** [comment]
Mate, I just spotted it's been you who provided the "stp-mt" submission in starexec:
https://www.starexec.org/starexec/secure/details/solver.jsp?id=23397

Do you update that solver, or shall I do that, and submit that to the organizers with the same name one more time?

I'll post a PR for the description within the next minutes. In case there are no big concerns, I'm sending that to the organizers tomorrow.

**Comment by msoos** [comment]
Sure, please update, let me try to give you access :)

**Comment by msoos** [comment]
OK, now both of you have access to my SMT space at: https://www.starexec.org/starexec/secure/explore/spaces.jsp?id=37559 

So both trev_abroad@yahoo.com and norbert.manthey@tu-dresden.de should now both be on the list of users. Let me know if this helps and thank you so much!

**Comment by conp-solutions** [comment]
I submitted another stp-mt solver, and will write an email to the organizers now.
https://www.starexec.org/starexec/secure/details/solver.jsp?id=23663

**Comment by msoos** [comment]
Thanks!!

Mate

**Comment by TrevorHansen** [comment]
SMTLIB has a single benchmark that needs 3-arity addition, which is now
allowed. It's just in the incremental track. I've pushed a hack to fix it.
Up to you guys whether to resubmit. Trev.

**Comment by msoos** [comment]
Sure, I'll resubmit in a few minutes, thank you!

Mate
